// all-in-one-pacakge/math.js

function add(a, b) { return a + b; }
function subtract(a, b) { return a - b; }
function multiply(a, b) { return a * b; }
function divide(a, b) { return b === 0 ? null : a / b; }

function power(a, b) { return a ** b; }
function factorial(n) {
  if (n < 0) return null;
  if (n === 0) return 1;
  let f = 1;
  for (let i = 1; i <= n; i++) f *= i;
  return f;
}

function coulombForce(q1, q2, r) {
  const k = 9e9;
  return (k * q1 * q2) / (r * r);
}


function gravityForce(m1, m2, r) {
  const G = 6.626e-11;
  return (G * m1 * m2) / (r * r);
}


function electricField(q, r) {
  const k = 9e9;
  return k * q / (r * r);
}

function potentialEnergy(m, g, h) {
  return m * g * h;
}


function kineticEnergy(m, v) {
  return 0.5 * m * v * v;
}


function magnitude(vector) {
  return Math.sqrt(vector.reduce((sum, val) => sum + val*val, 0));
}

module.exports = {
  add, subtract, multiply, divide, power, factorial,
  coulombForce, gravityForce, electricField,
  potentialEnergy, kineticEnergy, magnitude
};
